plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+}
+
+/* Issue a nvm auth write operation */
+static gboolean
+fu_dell_dock_usb4_hub_nvm_authenticate (FuDevice *device, GError **error)
+{
+	guint8 op_status;
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_AUTH_WRITE,
+					       NULL, &op_status, error)) {
+		g_prefix_error (error,
+				"Note: Succesful NVM Auth operation may cause "
+				"immediate GR reset? and thus problems to read "
+				"NVM Auth operation status");
+		return FALSE;
+	}
+	switch (op_status) {
+		case 0:
+			g_prefix_error (error,
+				  "NVM Auth write operation successful");
+			break;
+		case 1:
+			g_prefix_error (error,
+				  "NVM Auth write operation error");
+			break;
+		case 2:
+			g_prefix_error (error,
+				  "NVM Auth operation status: Retry NVM write");
+			break;
+		default:
+			g_prefix_error (error,
+				  "NVM Auth operation status unknown");
+			break;
+	}
+	/* FIXME pass op_status to caller in some way */
+	return FALSE;
+}
+
+static gboolean
+fu_dell_dock_usb4_activate (FuDevice *device, GError **error)
+{
+	return fu_dell_dock_usb4_hub_nvm_authenticate (device, error);
+}
+
+static const gchar*
+fu_dell_dock_usb4_show_nvm_info (FuDevice *device, GError **error)
+{
+	gchar *nvm_version;
+	guchar buf[64];
+	guint len;
+
+	len = 64;
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, buf, len, 0, error)) {
+		g_prefix_error (error, "NVM Version Read Failed");
+		return NULL;
+	}
+	nvm_version = g_strdup_printf ("%02x.%02x", buf[10], buf[9]);
+	return nvm_version;
+}
+
+static gboolean
+fu_dell_dock_usb4_write_fw (FuDevice *device, FuFirmware *firmware,
+			    FwupdInstallFlags flags, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	guint32 header_offset = 0;
+	gsize image_size = 0;
+	const guint8 *buffer;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	g_autofree gchar *dynamic_version    = NULL;
+	g_autofree gchar *dynamic_vendor_id  = NULL;
+	g_autofree gchar *dynamic_product_id = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	g_return_val_if_fail (device != NULL, FALSE);
+	g_return_val_if_fail (FU_IS_FIRMWARE (firmware), FALSE);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	buffer = g_bytes_get_data (fw, &image_size);
+	g_debug ("Total Image size: %" G_GSIZE_FORMAT, image_size);
+
+	/* get image version */
+	dynamic_version = g_strdup_printf ("%02x.%02x",
+					   buffer[self->blob_major_offset],
+					   buffer[self->blob_minor_offset]);
+	g_debug ("writing Thunderbolt firmware version %s", dynamic_version);
+
+	/* get image vendor-id */
+	dynamic_vendor_id = g_strdup_printf ("0x%02x%02x",
+					     buffer[self->vid_offset+1],
+					     buffer[self->vid_offset]);
+	g_debug ("writing Thunderbolt vendor-id %s", dynamic_vendor_id);
+
+	/* get image product-id */
+	dynamic_product_id = g_strdup_printf ("0x%02x%02x",
+					       buffer[self->pid_offset+1],
+					       buffer[self->pid_offset]);
+	g_debug ("writing Thunderbolt product-id %s", dynamic_product_id);
```

Comment:
I think you want to have a check here most likely that product id and vendor id match no?

Reply:
Yes, in a discussion with Intel if possible to get vid/pid from TBT commands in order to compare before writing.

Reply:
I think you already have a `fu_dell_dock_usb4_hub_nvm_read` function, it should just be using this to get the right offsets.

And if you do this successfully then it should be used in **probe** to dynamically make the GUID `TBT-********` that matches the one thunderbolt plugin makes for de-duplication purposes.

Reply:
> And if you do this successfully then it should be used in probe to dynamically make the GUID TBT-******** that matches the one thunderbolt plugin makes for de-duplication purposes.
I guess getting the vid/pid from NVM is important to determine moving out this plugin.